lib/incoming_mail/handler.rb

```diff
@@ -60,47 +60,24 @@ def forward_msg(msg)
     end
 
     def create_comment(msg)
-      proposal = find_proposal(find_public_id(msg)) or return
-      comment_text = find_comment_text(msg)
-      comment_user = find_comment_user(msg)
-      comment = nil
-      if proposal.has_subscriber?(comment_user)
-        # already in the loop, just add comment.
-        comment = proposal.comments.create(comment_text: comment_text, user: comment_user)
-      else
-        # yes, user adds self as observer, which also generates comment
-        proposal.add_observer(comment_user, comment_user, comment_text)
-        comment = proposal.comments.last
-      end
-      Dispatcher.on_comment_created(comment) # sends email
-      comment
-    end
-
-    def find_proposal(public_id)
-      Proposal.find_by_public_id(public_id) || Proposal.find(public_id)
-    end
-
-    def find_public_id(msg)
-      references = msg['headers']['References']
-      ref_re = /<proposal-(\d+)\@.+?>/
-      sbj_re = /Request\ #?([\w\-]+)/
+      # IMPORTANT that we check/add as observer before we create comment,
+      # since comment will create as a user if not already,
+      # and we want the reason logged.
+      parsed_email = InboundMailParser.new(msg)
+      proposal = parsed_email.proposal
+      comment_user = parsed_email.comment_user
 
-      if references.match(ref_re)
-        references.match(ref_re)[1]
-      elsif msg['subject'].match(sbj_re)
-        msg['subject'].match(sbj_re)[1]
-      else
-        fail "Failed to find public_id in msg #{msg.inspect}"
+      unless proposal.has_subscriber?(comment_user)
+        add_user_as_observer(comment_user, proposal)
       end
-    end
 
-    def find_comment_text(msg)
-      EmailReplyParser.parse_reply(msg['text'])
+      comment = CommentCreator.new(parsed_email).run
+      Dispatcher.on_comment_created(comment) # sends email
+      comment
     end
 
-    def find_comment_user(msg)
-      from = msg['from_email']
-      User.find_by_email_address(from)
+    def add_user_as_observer(user, proposal)
+      proposal.add_observer(user, user, "Added comment via email reply")
```

Comment:
My thinking was that the commenter (user) was adding themselves by virtue of creating a comment. In practice I don't expect this to happen often, since someone would have to respond to an email related to a request for which they were not already a subscriber. A chicken and egg dilemma.